src/components/RegistrationForm/RegistrationForm.tsx

```diff
@@ -0,0 +1,198 @@
+import { useState, FormEvent, ChangeEvent, useMemo } from 'react';
+import { Link } from 'react-router-dom';
+
+import { FormInput } from '../FormInput/FormInput';
+import style from './RegistrationForm.module.scss';
+
+type FormData = {
+  firstName: string;
+  lastName: string;
+  email: string;
+  password: string;
+  confirmPassword: string;
+};
+
+type TouchedFields = {
+  firstName: boolean;
+  lastName: boolean;
+  email: boolean;
+  password: boolean;
+  confirmPassword: boolean;
+};
+
+const INITIAL_FORM_DATA: FormData = {
+  firstName: '',
+  lastName: '',
+  email: '',
+  password: '',
+  confirmPassword: '',
+};
+
+const INITIAL_TOUCHED: TouchedFields = {
+  firstName: false,
+  lastName: false,
+  email: false,
+  password: false,
+  confirmPassword: false,
+};
+
+const ERROR_MESSAGES = {
+  firstName: 'First name cannot be empty or contain numbers',
+  lastName: 'Last name cannot be empty or contain numbers',
+  email: 'Please enter a valid email',
+  password:
+    'Password must have at least 8 characters, 1 uppercase letter and 1 number',
+  passwordMatch: "Passwords don't match",
+};
+
+const validators = {
+  name: (value: string): boolean => /^[^0-9]+$/.test(value) && value.length > 0,
+  email: (value: string): boolean => /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(value),
+  password: (value: string): boolean =>
+    /^(?=.*[A-Z])(?=.*\d).{8,}$/.test(value),
+};
+
+export const RegistrationForm = () => {
+  const [formData, setFormData] = useState<FormData>(INITIAL_FORM_DATA);
+  const [touched, setTouched] = useState<TouchedFields>(INITIAL_TOUCHED);
+  const [shakeAnimation, setShakeAnimation] = useState(false);
+  const [isSubmitted, setIsSubmitted] = useState(false);
+
+  const formErrors = useMemo(
+    () => ({
+      firstName: touched.firstName && !validators.name(formData.firstName),
+      lastName: touched.lastName && !validators.name(formData.lastName),
+      email: touched.email && !validators.email(formData.email),
+      password: touched.password && !validators.password(formData.password),
+      confirmPassword:
+        touched.confirmPassword &&
+        formData.password !== formData.confirmPassword,
+    }),
+    [formData, touched],
+  );
```

Comment:
This validates the whole form, on every change. I believe this should not sit in useMemo but in the handleChange, so we can drop the touching.

We will need to add another state. But that is fine. You can still use the validators abstraction